rama-cli/src/cmd/tls/mod.rs

```diff
@@ -0,0 +1,152 @@
+use clap::Args;
+use rama::{
+    Context, Layer, Service,
+    error::BoxError,
+    net::{
+        address::Authority,
+        client::{ConnectorService, EstablishedClientConnection},
+        tls::{
+            DataEncoding,
+            client::{ClientConfig, NegotiatedTlsParameters, ServerVerifyMode},
+        },
+    },
+    tcp::client::{Request, service::TcpConnector},
+    tls::{
+        rustls::client::{TlsConnectorData, TlsConnectorLayer},
+        std::dep::boring::x509::X509,
+    },
+};
+use tokio::net::TcpStream;
+use tracing::level_filters::LevelFilter;
+use tracing_subscriber::{EnvFilter, fmt, layer::SubscriberExt, util::SubscriberInitExt};
+
+#[derive(Args, Debug, Clone)]
+/// rama tls support
+pub struct CliCommandTls {
+    /// The address to connect to
+    /// e.g. "example.com" or "example.com:8443"
+    /// if no port is provided, the default port 443 will be used
+    address: String,
+
+    #[arg(long, short = 'k')]
+    /// Wether to skip certificate verification
+    insecure: bool,
+}
+
+/// Run the tls command
+pub async fn run(cfg: CliCommandTls) -> Result<(), BoxError> {
+    tracing_subscriber::registry()
+        .with(fmt::layer())
+        .with(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env_lossy(),
+        )
+        .init();
+
+    let authority = match Authority::try_from(cfg.address.clone()) {
+        Ok(authority) => authority,
+        // if missing port, we can try to add the default port
+        Err(err) if err.to_string().contains("missing port") => {
```

Comment:
First, never rely on error strings. If you do ever need to match on an error we would need to have a specific error type you can check for via BoxError/stdError.

However I think you can keep it a lot more simpler. Something like:

```
let address = cfg.address.trim();
let authority: Authority = if cfg.address.contains(':') {
     address.parse().context("parse config address as authority")?
} else {
     let host: Host = address.parse().context("parse config address as host")?;
     Authority::new(host, 443)
};
```

Doesn't that do exactly what you want without having to rely on some specific errors even?

Reply:
Thank you, this is much better indeed, learned something new ;)